pkg/istiod/common.go

```diff
@@ -0,0 +1,324 @@
+		},
+		Config: ConfigArgs{},
+
+		// MCP is messing up with the grpc settings...
+		MCPMaxMessageSize:        1024 * 1024 * 64,
+		MCPInitialWindowSize:     1024 * 1024 * 64,
+		MCPInitialConnWindowSize: 1024 * 1024 * 64,
+	}
+
+	// Main server - pilot, registries
+	server, err := NewServer(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := server.WatchMeshConfig(meshCfgFile); err != nil {
+		return nil, fmt.Errorf("mesh: %v", err)
+	}
+
+	pilotAddress := server.Mesh.DefaultConfig.DiscoveryAddress
+	_, port, _ := net.SplitHostPort(pilotAddress)
+	basePortI, _ := strconv.Atoi(port)
+	basePortI = basePortI - basePortI % 100
+	basePort := int32(basePortI)
+	server.basePort = basePort
+
+
+	args.DiscoveryOptions =  envoy.DiscoveryServiceOptions{
+		HTTPAddr:        fmt.Sprintf(":%d", basePort+7),
+		GrpcAddr:        fmt.Sprintf(":%d", basePort+10),
+		// Using 12 for K8S-DNS based cert.
+		// TODO: We'll also need 11 for Citadel-based cert
+		SecureGrpcAddr:        fmt.Sprintf(":%d", basePort+12),
+		EnableCaching:   true,
+		EnableProfiling: true,
+	}
+	args.CtrlZOptions = &ctrlz.Options{
+		Address: "localhost",
+		Port:    uint16(basePort + 13),
+	}
+
+		err = server.InitConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Galley args
+	gargs := settings.DefaultArgs()
+
+	// Default dir.
+	// If not set, will attempt to use K8S.
+	gargs.ConfigPath = baseDir + "/var/lib/istio/local"
+	// TODO: load a json file to override defaults (for all components)
+
+	gargs.ValidationArgs.EnableValidation = false
+	gargs.ValidationArgs.EnableReconcileWebhookConfiguration = false
+	gargs.APIAddress = fmt.Sprintf("tcp://0.0.0.0:%d", basePort+901)
+	gargs.Insecure = true
+	gargs.EnableServer = true
+	gargs.DisableResourceReadyCheck = true
+	// Use Galley Ctrlz for all services.
+	gargs.IntrospectionOptions.Port = uint16(basePort + 876)
+
+	// The file is loaded and watched by Galley using galley/pkg/meshconfig watcher/reader
+	// Current code in galley doesn't expose it - we'll use 2 Caches instead.
+
+	// Defaults are from pkg/config/mesh
+
+	// Actual files are loaded by galley/pkg/src/fs, which recursively loads .yaml and .yml files
+	// The files are suing YAMLToJSON, but interpret Kind, APIVersion
+
+	gargs.MeshConfigFile = meshCfgFile
+	gargs.MonitoringPort = uint(basePort + 15)
+	// Galley component
+	// TODO: runs under same gRPC port.
+	server.Galley = NewGalleyServer(gargs)
+
+	// TODO: start injection (only for K8S variant)
+
+	// TODO: start envoy only if TLS certs exist (or bootstrap token and SDS server address is configured)
+	//err = startEnvoy(baseDir, &mcfg)
+	//if err != nil {
+	//	return err
+	//}
+	return server, nil
+}
+
+func (s *Server) WaitDrain(baseDir string) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	<-sigs
+	// Will gradually terminate connections to Pilot
+	DrainEnvoy(baseDir, s.Args.MeshConfig.DefaultConfig)
+
+}
+
+
+var trustDomain = "cluster.local"
```

Comment:
doesn't look like this is used, but if it is it should come from mesh config instead

Reply:
Done.